### ISSUE TITLE
time_stats forms inappropriate URI

### ISSUE THREAD

**Issue by ebergerson** [python-gitlab/python-gitlab#510]
## Description of the problem, including code/CLI snippet

I have used the api to grab a list of issues and iterate through them getting their time stats:

    def display_issues(self, issues: Iterator[ProjectIssue]):
        rows = []
        for issue in issues:
            row = {k: getattr(issue, k) for k in ['iid', 'title', 'state', 'closed_at']}
            row['closed_at'] = pd.to_datetime(row['closed_at'], utc=True).strftime("%Y-%m-%d %H:%M") \
                if row['closed_at'] else ""
            ts = issue.time_stats()
            row['estimate'] = ts['human_time_estimate']
            row['spent'] = ts['human_total_time_spent']
            row['assignee'] = issue.assignee['username'] if issue.assignee else ""
            rows.append(row)

It takes a very long time.  I discovered when running in debug mode that each call to time_stats() is failing once on an inappropriate formed URI.  Here is the debugging output

DEBUG:root:Stage: Display Issues
DEBUG:urllib3.connectionpool:https://gitlab.com:443 "GET /api/v4/projects/5118242/issues//237/time_stats HTTP/1.1" 308 0
DEBUG:urllib3.connectionpool:Resetting dropped connection: gitlab.com
DEBUG:urllib3.connectionpool:https://gitlab.com:443 "GET /api/v4/projects/5118242/issues/237/time_stats HTTP/1.1" 200 105
DEBUG:urllib3.connectionpool:https://gitlab.com:443 "GET /api/v4/projects/5118242/issues//235/time_stats HTTP/1.1" 308 0
DEBUG:urllib3.connectionpool:Resetting dropped connection: gitlab.com
DEBUG:urllib3.connectionpool:https://gitlab.com:443 "GET /api/v4/projects/5118242/issues/235/time_stats HTTP/1.1" 200 105
DEBUG:urllib3.connectionpool:https://gitlab.com:443 "GET /api/v4/projects/5118242/issues//227/time_stats HTTP/1.1" 308 0
DEBUG:urllib3.connectionpool:Resetting dropped connection: gitlab.com
DEBUG:urllib3.connectionpool:https://gitlab.com:443 "GET /api/v4/projects/5118242/issues/227/time_stats HTTP/1.1" 200 97
DEBUG:urllib3.connectionpool:https://gitlab.com:443 "GET /api/v4/projects/5118242/issues//226/time_stats HTTP/1.1" 308 0
DEBUG:urllib3.connectionpool:Resetting dropped connection: gitlab.com
....

I have found the error is in this code:

class TimeTrackingMixin(object):
    @cli.register_custom_action(('ProjectIssue', 'ProjectMergeRequest'))
    @exc.on_http_error(exc.GitlabTimeTrackingError)
    def time_stats(self, **kwargs):
        """Get time stats for the object.

        Args:
            **kwargs: Extra options to send to the server (e.g. sudo)

        Raises:
            GitlabAuthenticationError: If authentication is not correct
            GitlabTimeTrackingError: If the time tracking update cannot be done
        """
        path = '%s/%s/time_stats' % (self.manager.path, self.get_id())
        return self.manager.gitlab.http_get(path, **kwargs)

Setting a breakpoint before the return we can see that the path is ill formed:

![image](https://user-images.githubusercontent.com/1410507/40735199-91f8e03e-6408-11e8-8622-c6151c945ce7.png)

The problem is that the self.manager.path has a terminating `/`.

![image](https://user-images.githubusercontent.com/1410507/40735239-aa2610be-6408-11e8-9757-5487bfd88578.png)

## Specifications

  - python-gitlab version: 1.4.0
  - API version you are using (v3/v4): V4
  - Gitlab server version (or gitlab.com): GitLab.com

**Comment by ebergerson** [comment]
BTW, I tried to figure out how the URL gets corrected for the subsequent successful call for each time_stat and could not find it.   However, this runs VERY slow.  Is there a way to query issues with time_stats pre-populated since I need it for every issue?

**Comment by ebergerson** [comment]
Ok, so I just discovered that the issue class has an _attrs member that contains a dictionary with the time_stats already loaded when the issue is first pulled down.

```
issue._attrs['time_stats']
{'time_estimate': 14400, 'total_time_spent': 14400, 'human_time_estimate': '4h', 'human_total_time_spent': '4h'}
```
However, calling issue.time_stat() performs a new request, the one that is indicated as broken above:

```
issue.time_stats()
DEBUG:urllib3.connectionpool:Resetting dropped connection: gitlab.com
DEBUG:urllib3.connectionpool:https://gitlab.com:443 "GET /api/v4/projects/5118242/issues//235/time_stats HTTP/1.1" 308 0
DEBUG:urllib3.connectionpool:Resetting dropped connection: gitlab.com
{'time_estimate': 14400, 'total_time_spent': 14400, 'human_time_estimate': '4h', 'human_total_time_spent': '4h'}
DEBUG:urllib3.connectionpool:https://gitlab.com:443 "GET /api/v4/projects/5118242/issues/235/time_stats HTTP/1.1" 200 105
```

I hate breaking 'encapsulation' and directly accessing an underscore member of a class.  What is the appropriate way to access the loaded time stats without using the method that makes another request?

**Comment by gpocentek** [comment]
Hi @ebergerson 

There is indeed an issue in the path. gitlab responds to the first query with a 308 redirect, and the 2nd query is the "real" one. But it seems that gitlab.com takes time to answer the request, even with the correct path (ended up with timeouts while testing). There's nothing I can do to fix that.

I don't think there is a way to get the time stats pre-populated, or it is not documented.

**Comment by ebergerson** [comment]
@gpocentek Thank you so much for your reply.  Look at my last comment, I indeed did find the pre-populated data.  It was hinted at in the [GitLab API doc](https://docs.gitlab.com/ee/api/issues.html), since it is in the returned data structure.  

Do you know if there is a better way to access the data than going through the private member?

**Comment by ebergerson** [comment]
Maybe the method time_stats() should be changed to simply return the internal data.  Should not break anyone's code and be faster for all!

**Comment by gpocentek** [comment]
@ebergerson you can access the attribute using `issue.attributes['time_stats']`

Thanks for your research :) I'll check if I can update the `time_stats()` method to handle older versions of GitLab which didn't return the info in the issue objects.

**Comment by ebergerson** [comment]
Yeah, should be pretty easy to check if the data is local and only go for a new request if it is not already present.  Thank you.
